Clustermesh APIserver: Add CFP to filter ciliumendpoints, identities and endpointslices exported to etcd (#74)
cilium/design-cfps: cilium/CFP-39876-clustermesh-filtered-export.md

# CFP-39876: Scoped Export Mode for ClusterMesh

**SIG:** SIG-NAME

**Sharing:** Public

**Begin Design Discussion:** 2025-06-02

**End Design Discussion:**

> **Cilium Release:** (TBD)

Reviewer: I assume that the target would be Cilium v1.19. Is that correct?

**Authors:** Krunal Jain <krunaljain@microsoft.com>, Vamsi Kalapala <Krishna.Vamsi@microsoft.com>

## Summary

> This CFP introduces selective distribution of endpoints and identities within Cilium Clustermesh by introducing a new “scoped-export” mode for the Clustermesh API server, restricting cross-cluster propagation only to resources that are fronted by global services. The modification targets improved scalability while acknowledging that direct inter-cluster endpoint access and network policy enforcement will be discontinued for resources not associated with global services. The addition of new Clustermesh APIserver export mode would ensure the changes are backwards compatible.

Reviewer: > restricting cross-cluster propagation only to resources that are fronted by global services

From a first read, I'm not really convinced by this approach, as I personally feel it would not be intuitive from a user perspective. The concept of service in Kubernetes (and by extension Cluster Mesh global service) clearly maps to a set of destination pods abstracted by a single entity. However, a (global) service can be accessed by any client in the cluster, which is not expected to be behind a service (unless it is also playing the backend role). In turn, this would mean that users may easily end up with a pod to service flow, which is explicitly called out as a non-goal as part of this proposal, with unexpected consequences. 

Did you consider having the export scopes correspond to namespace boundaries, allowing users to annotate specific namespaces to elect them as part of the clustermesh. To me, that feels way more intuitive from the user perspective, and would similarly solve the scalability concerns under the assumption that applications that need to communicate across clusters are scoped to dedicated namespaces. It would be also fairly trivial to warn users via admission webhooks if they mark a service as global (or create a ServiceExport resource) in a non-global namespace, if that's desired.

From an implementation point of view, it would be also mostly trivial to extend the clustermesh-apiserver to filter the synchronization of CiliumEndpoints/CiliumEndpointSlices and CiliumIdentities by namespace, and way less prone to possible race conditions and surprising behavior.

Reviewer 2: I think this is the most important part to establish for this CFP. Other parts are mostly implementation details. 

> This CFP introduces selective distribution of endpoints and identities within Cilium Clustermesh by introducing a new “scoped-export” mode for the Clustermesh API server, restricting cross-cluster propagation only to resources that are fronted by global services.

To extend a little bit on Marco's concern and explain it in a bit easier terms. 
If client A tries to connect to global service S with a backend B in another cluster:
```
Pod A running on node A (Cluster1) -> Global Service S -> Pod B running on node B (Cluster2) 
```
For service translation, we need information about global service and its backends.
You can think of it that after service translation is performed, it's the same case as regular pod A to pod B connectivity.

However, for pod A to pod B connectivity to work, we need information about the endpoint, identity, and corresponding node information on both ends **in general**.
For pod A - endpoint/identity information of pod B is used for enforcing the Egress network policy.
For pod B - endpoint/identity information of pod A is used for enforcing the Ingress network policy.

There might be different deployment methods that theoretically would not require client pod A information to be propagated through Clustermesh to other clusters - in the case of tunneling, as the tunnel itself contains identity information. However, I would consider that as a potential improvement over the base case, which requires client pod information to be propagated.
For native routing, the only information that we get from the connection is IP, which we need to somehow map to identity to enforce the Ingress network policy on the backend. That maybe could work without identity information if the backend would allow for "world" Ingress traffic for example, which might be a security concern on the other hand. 

So to sum up, I think for the "base" case, we need to propagate the endpoint and identity information of both **clients** and **global service backends** across clusters.
Now the most important design question is how would you like it to work. 
Some ideas I have and each of them has it's own pros and cons:
- One proposal from Marco is that we annotate namespaces and only pods within an annotated namespace can access global services.
- Alternatively, each client pod is annotated which allows for global-service access and we propagate information about all global services.
- Maybe some other idea - for example with proxy pods that are only allowed to communicate across clusters and all connections are proxied through them. 
- Or go with something related to hybrid mode in https://github.com/cilium/design-cfps/pull/73 where cross-cluster traffic is always tunneled and then we don't need endpoint/identity information of the clients to enforce policies.

Each of the proposals have their own pros and cons, which I would like to explore first before deciding on the design.

Reviewer: >>  Did you consider having the export scopes correspond to namespace boundaries, allowing users to annotate specific namespaces to elect them as part of the clustermesh.

> Maybe this could be an iterative process?

Yep, absolutely. I'm not opposed to the other approaches in principle, but I would start with an intuitive (and simple to implement) approach at the beginning, to cover the vast majority of use-cases. Then it is always possible to iterate if necessary to cover more advanced use cases.

Author: @marseel  @giorio94 Updated the CFP highlighting network policy support only for CiliumIdentities in the allowlisted namespaces. This would imply that pods that are relevant to the policy are always explicitly enabled for export by customers

> This CFP introduces selective distribution of endpoints and identities within Cilium Clustermesh by introducing a new “scoped-export” mode for the Clustermesh API server, restricting cross-cluster propagation only to resources that are fronted by global services. The modification targets improved scalability while acknowledging that direct inter-cluster endpoint access and network policy enforcement will be discontinued for resources not associated with global services. The addition of new Clustermesh APIserver export mode would ensure the changes are backwards compatible.

Reviewer: > will be discontinued

I'd suggest clarifying what *discontinued* means in practice.

Author: Based on the feedback to allowlist pods for export, we can support network policies as is. I will remove this part.

Author: Going with the namespace allowlisting approach. Updated the goals for Network Policy support

## Motivation

The existing Cilium Clustermesh implementation distributes all endpoints and identities across connected clusters, creating scaling constraints as resource volumes grow. This comprehensive propagation model becomes a limiting factor when managing extensive endpoint and identity collections. Performance profiling analysis demonstrates the severity of this scaling challenge: in a 200-cluster Clustermesh deployment with the following configuration parameters, each Cilium agent consumes 15GiB of memory:

- cm-apiserver-replicas=2
- clusters=200
- nodes/cluster=300
- identities/cluster=500
- identities-qps=0.2

> - endpoints/cluster=15000

Author: Noted. The numbers were used to demonstrate the scale limitations but since they fall beyond the support range, I will remove them.

- endpoints-qps=1
- services=0
- services-qps=0

> By restricting distribution to only those resources backing global services, we can achieve substantial scalability improvements and operational efficiency gains. The gains in compute from the proposed optimization would also ensure that customers can run multiple replicas of the Clustermesh APIserver which is recommended for production workloads.

Author: @giorio94 The identity and endpoint volume exported to etcd increases in multiples as we add multiple CM APIserver replicas. This does not scale well with high volume. The proposed optimization is expected to reduce export volume and in turn enabling customers to run multiple replicas for their production workloads if required. I will expand this section a bit to make this clear.

> By restricting distribution to only those resources backing global services, we can achieve substantial scalability improvements and operational efficiency gains. The gains in compute from the proposed optimization would also ensure that customers can run multiple replicas of the Clustermesh APIserver which is recommended for production workloads.

Reviewer: > The identity and endpoint volume exported to etcd increases in multiples as we add multiple CM APIserver replicas. This does not scale well with high volume

That's not how the clustermesh-apiserver works. Every clustermesh-apiserver replica synchronizes exactly the same amount of information into its own sidecar etcd instance, and every etcd instance is completely independent of all the others (regardless of the optimization that you are proposing).

> By restricting distribution to only those resources backing global services, we can achieve substantial scalability improvements and operational efficiency gains. The gains in compute from the proposed optimization would also ensure that customers can run multiple replicas of the Clustermesh APIserver which is recommended for production workloads.

Reviewer: > The blocker is more from the compute side rather than the functionality. For 2 replicas, the work required for Clustermesh APIserver to sync remote identities and endpoints will be doubled. For a high identity and endpoint volume clustermesh, this can result in customers running out of compute resources to be able to run 2 replicas.

OK, I think I'm understanding a bit better your point. 

Based on the scale testing we perform in Cilium (~1 order of magnitude lower than your mentioned scale), the clustermesh-apiserver (including both the clustermesh and kvstoremesh containers) is not particularly resource hungry, especially in terms of CPU (< 1 core). I wouldn't expect that to change much based on the scale, as mostly throttled by the rate limiter settings, unless you significantly increased them. Memory is likely a different story, as it grows proportionally to the amount of ingested resources. In any case, most of the resource usage comes from the etcd watchers, and increasing the number of replicas allows to distribute the load, so that gets compensated.

> The network bandwidth to transmit the endpoints and identities across the clusters will also be 2x.

Yes, due to how kvstoremesh works. That said, any agent needs to pull and ingest the same amount of information, so to me most of the benefit comes from there, compared to the actual reduction in clustermesh-apiserver resources.

## Goals

- Enhance Cilium Clustermesh scalability through selective endpoint and identity distribution
- Maintain cross-cluster visibility for endpoints and identities associated with global services
- Service to Service connection and network policies

> ## Non-Goals

Author: Yes this is supported only with CRD mode. Will explicitly mention this in the CFP.

- Direct inter-cluster endpoint access for resources not backed by global services

> - Cross-cluster network policy enforcement

Author: > * Is all non-explicitly allowed cross-cluster traffic expected to be dropped if any network policy is applied?
Yes. If a deny by default policy is applied, traffic would only be allowed if 
1) The pod's identity is exported to remote 
2) A policy explicitly allows traffic from that cluster
> * Which of the existing selectors, if any, would allow traffic from/to pods in remote clusters? Would it be perceived as `world` traffic?
If the remote pod's identity is not exported, I believe it would be perceived as world traffic and would be dropped

> - Cross-cluster network policy enforcement

Reviewer: >> Is all non-explicitly allowed cross-cluster traffic expected to be dropped if any network policy is applied?
>
> Yes. If a deny by default policy is applied, traffic would only be allowed if
> 1. The pod's identity is exported to remote
> 2. A policy explicitly allows traffic from that cluster

If this is the expectation, then it should be explicitly codified as part of the CFP. 

However, based on your comment below (`[...] I believe it would be perceived as world traffic [...]`), traffic would also be allowed if e.g., `to-world` traffic is allowed (I'm not totally sure if there's an equivalent `from-world` for ingress policies, but that would apply as well). I'm not saying that would not be acceptable, but again it needs to be explicitly stated if that is the expectation, to get agreement on it.

>> Which of the existing selectors, if any, would allow traffic from/to pods in remote clusters? Would it be perceived as world traffic?
>
> If the remote pod's identity is not exported, I believe it would be perceived as world traffic and would be dropped

I don't think that would be the case in all scenarios, and that's why I had mentioned the different possibilities in my initial comment. Most notably, as Marcel mentioned in the comment below, the identity is carried inside the tunnel when Cilium runs in encapsulation mode. So, considering the:

> Source pod not exported, destination pod not exported, ingress policy

case, Cilium on the destination node would see the actual source pod identity (carried via the tunnel) --- not `world` ---, and eventually discard the traffic as not allowed by the ingress policy. Differently, in native routing mode, it would try looking up the mapping into the ipcache, and fall back to the world identity given that no explicit mapping is actually found. The policy may then either allow or discard the traffic depending on whether `world` is accepted. Hence, the behavior would be inconsistent and different depending on the routing mode, which is suboptimal.

Another special case is represented by IPAM modes associating one or more PodCIDRs to any given node. Then, corresponding entries would be inserted into the ipcache, and would be matched for pods that are not shared (rather than matching the catch-all world entry). Currently, these entries are also associated with the `world` identity for historical reasons, but that may change in the future. That's again something to keep in mind in this context.

One more case to consider, which I have not reasoned about yet, is how this feature would interact with other types of policies, such as `toCIDR`, given it somewhat changes the normal expectations.

- Pod to Pod or Pod to service connection or Network Policy

## Proposal

### Implementation details

> We intend to add a new cache in conjunction to the Service cache that exists both in the cilium agent as well as the operator. We intend to provide a new config in the cilium agent configmap called scoped-export which is by default set to false. Upon setting this config, the agent and the operator will annotate the CiliumEndpoint, CiliumEndpointSlice and CiliumIdentity CRDs with internal annotation allowlisting them for export to etcd. A change in scope–export config would require restart of cilium agent and operator components for the changes to take effect.

Reviewer: A couple of high-level comments here:
* Did you consider performing the filtering entirely within the clustermesh-apiserver, rather than having Cilium agents/operator take care of annotating the resources in advance? Any reason why you preferred the current approach?
* The concept of `Service cache` no longer strictly exists \[1], as replaced by the new load-balancing control-plane logic.
* Services can be marked as global both via the dedicated annotation and creating a corresponding `ServiceExport` resource, if MCS-API support is enabled.

\[1]: https://github.com/cilium/cilium/pull/40022

Reviewer 3: > Did you consider performing the filtering entirely within the clustermesh-apiserver, rather than having Cilium agents/operator take care of annotating the resources in advance? Any reason why you preferred the current approach?

FYI this was the original proposal but this seems a bit counter intuitive to me because as CEP/CES/Identities are tied to a/some pods and the original proposal considered this only for global service with possible pod/namespace annotation being only a future consideration not included at first.

So I was essentially suggesting to make something around pod annotation be the primary way of doing that rather than service with an option that the service could even auto add annotation on existing pods so that you can reuse most of the implementation/code (although I am not fully convinced if we were to add that if the annotation should be added on CEP directly or the pod).

That being said I didn't really consider the namespace approach at that time and I agree with your other message suggesting that namespace are a good idea and focusing on that at least at first would makes a lot of sense!

If the operator manages CiliumIdentities, it responds to global service events from its cache and performs reconciliation to add or remove internal annotation as needed.

#### CiliumEndpointSlice Controller Updates

The controller, which already watches CiliumEndpoint resources, is updated to check for the global backend annotation. If present, it adds a corresponding annotation to the CiliumEndpointSlice. Annotation changes are tracked and updated accordingly.

### ClusterMesh API Server Changes

#### Export Filtering

The ClusterMesh API server filters resources before exporting to etcd. Only those with the internal global backend annotation are shared for cross-cluster visibility.

#### Supported Resource Types

The filter applies to CiliumEndpoint, CiliumIdentity, and CiliumEndpointSlice. Only annotated resources are exported to be shared cross clusters.

## Future Milestones

### Dynamic Export Scope Configuration

Introduce support for dynamic configuration of export scope per namespace or pod label. This would allow operators to fine-tune which endpoints and identities are shared across clusters beyond just global service association.

### Policy-Aware Export Filtering

> Integrate network policy awareness into the export decision logic. Endpoints and identities could be exported based on whether they are referenced in cross-cluster network policies, even if not fronted by a global service.

Reviewer: > Integrate network policy awareness into the export decision logic.

This is not necessarily possible, unless policy information is additionally exchanged cross-cluster. Consider for instance the following scenario:

* Client A1 hosted in cluster A
* Server B1 hosted in cluster B
* A1 wants to access B1 via a global service
* B1 is matched by an ingress policy to allow incoming traffic from A1

To make that work, cluster B needs to be aware of A1's IP to identity mapping (i.e., its CiliumEndpoint), at least in native routing mode, and A1's identity details (i.e., the CiliumIdentity). However, cluster A would have no ways of knowing that such a policy exists in B, and hence that these resources would need to be shared.

Reviewer: It should be integrated in this PR as well, as it serves as future reference.

| Yes                     | No                            | Egress Policy           | ❌ Works for tunnel mode. Does not work for native routing mode |
| No                      | No                            | Any Policy              | ❌ Works for tunnel mode. Does not work for native routing mode  |
| Any                     | Any                           | No Policies             | ✅ Allowed (no restrictions) |

Even though the policies work for tunnel routing mode, for consistency in customer experience, we do not provide support for network policies for workloads outside global namespaces. 

## Proposal

### Implementation details

We intend to only export ciliumendpoints and ciliumidentities for global namespaces when the clustermesh-apiserver runs in "scoped-export" mode. This new mode has three configs which the clustermesh-apiserver imports from cilium configmap namely 
- scoped-export-enabled - Enable the scoped export mode 
- global-namespaces - List of namespaces that are "allowlisted for export". The ciliumendpoints and ciliumidentities inside the allowlisted namespaces are exported to remote clusters. These are also known as global namespaces. 

> - allow-by-default - Allowlisting method. If set to true, all namespaces are disabled for export until explicitly exported using the allowlisted-namespaces configs. If set to false,  ciliumendpoints and identities under all namespaces are enabled for export by default until mentioned under the `global-namespaces`. 

Author: Yes I recommended the other approach that we discussed of having this list as a static config. The advantage of this approach are:
1) We don't need to have additional watch on namespaces on clustermesh-apiserver
2) Customers can easily allowlist/denylist namespaces and have a single point of reference instead of tracking which namespaces have/haven't been annotated

> - allow-by-default - Allowlisting method. If set to true, all namespaces are disabled for export until explicitly exported using the allowlisted-namespaces configs. If set to false,  ciliumendpoints and identities under all namespaces are enabled for export by default until mentioned under the `global-namespaces`. 

Reviewer 3: I am personally not that thrilled about this but let's see what @marseel or @giorio94 think about this

> - allow-by-default - Allowlisting method. If set to true, all namespaces are disabled for export until explicitly exported using the allowlisted-namespaces configs. If set to false,  ciliumendpoints and identities under all namespaces are enabled for export by default until mentioned under the `global-namespaces`. 

Reviewer 2: I would like to choose solution that is the most user-friendly.
Restarting clustermesh every time you want to add/delete global namespace sounds problematic from user perspective - regardless if it's "static" configmap or dynamic list of namespaces through annotations.
I understand that it might be a bit simpler to implement with static list, but I don't think there is a significant difference.

> We don't need to have additional watch on namespaces on clustermesh-apiserver

I don't see how does it matter?
I'm just asking as each cilium-agent already opens namespace informer, so one more doesn't matter that much at cluster level :sweat_smile: 

> Customers can easily allowlist/denylist namespaces

I don't think they are able to do it easily, they need more permissions, like restarting clustermesh, updating configmap vs annotating namespace. 

> have a single point of reference instead of tracking which namespaces have/haven't been annotated.

This is subjective, both can be achieved with a single kubectl command (either getting configmap or listing namespaces with specific annotation). 

> Updating namespace list requires a pod restart (Note: This would be required anyways since we don't have mechanism to backfill newly added global namespaces in etcd)

As mentioned above, I don't agree. 
I think implementing backfill mechanism should be fairly simple to achieve, without requiring pod restart. 
Configmaps will always require restarts, while namespace annotations won't.

However, If your particular use-case requires you to always have static list of namespaces, I wouldn't be super opposed to having two different sources of information which namespaces are exported to choose from - by default namespace annotation, and as an alternative, list of namespaces in configmap.

> - allow-by-default - Allowlisting method. If set to true, all namespaces are disabled for export until explicitly exported using the allowlisted-namespaces configs. If set to false,  ciliumendpoints and identities under all namespaces are enabled for export by default until mentioned under the `global-namespaces`. 

Author: @marseel @MrFreezeex I have updated the CFP to use annotations. Summarizing the changes made:
1) Add watch on namespaces to track annotations for allowlist/denylist. 
2) Trigger full sync when new namespaces annotated. Add/Remove existing entries from etcd not matching annotations 
I have removed dependencies on the configmap since that would've also required restarts. Let me know if this works.

## MCS support
Similar to global services, MCS support is only available for services created under global namespaces. 

## Implementation details

The proposed scoped-export mode has a config that users can pass in to mark all non local annotated namespaces as global. 
- scoped-export-default-enabled - true/false Honors only the local namespace annotation and marks all non annotated namespaces as global if set to true. Honors the global annotation and marks all non annotated  namespaces as local if set to true

#### Allow/Deny listing namespace
To allowlist a namespace and mark it as global, users can annotate the namespace with:

```
clustermesh.cilium.io/global: "true"
```

The ClusterMesh API server will watch namespace resources. Adding this annotation to any namespace will result in the following steps:

- A full synchronization is triggered for all resources within the newly annotated namespaces 
- Only entries associated with namespaces explicitly annotated as global will be exported.

> - Unannotated namespaces are marked global/local depending on the value of config `scoped-export-default-enable `

Reviewer 3: If we are maintaining the "global namespace" (vs an annotation having "export" in some form in its) I think the option should be named something like `clustermesh-default-global-namespace` for consistency with its annotation

### Behavior When Pods Are Not in Global Namespaces

If either the source or the destination pod does **not** belong to a global namespace, cross-cluster communication **may still succeed**, but under certain conditions although we do not support as part of the new mode. The evaluation of network policy for different cases is summarize below:

### Summary of Enforcement Logic

| Source Pod in Global NS | Destination Pod in Global NS | Network Policy Applies | Traffic Allowed? |
|-------------------------|-------------------------------|-------------------------|------------------|
| Yes                     | Yes                           | Yes                     | ✅ Enforced normally |
| No                      | Yes                           | Ingress Policy          | ❌ Not Supported  |
| Yes                     | No                            | Egress Policy           | ❌ Not Supported |
| No                      | No                            | Any Policy              | ❌ Not Supported  |
| Any                     | Any                           | No Policies             | ✅ Allowed (no restrictions) |

## Global Service Support

> In the namespace based export mode, all services within global namespaces would automatically be marked as global/local depending on the provided annotation on the namespace. Users are not required to explicitly annotate individual services. Any service which is not inside a global namespace is considered local even if it has an associated global annotation. Such a service will not share backends with remote clusters in any case.  

Author: @MrFreezeex  "I am not sure I would be in favor of doing this at all"
By this, you imply not doing the optimization that allows global by default or keeping the global services functionality as it is irrespective of if the service resides in global/local namespace?

> In the namespace based export mode, all services within global namespaces would automatically be marked as global/local depending on the provided annotation on the namespace. Users are not required to explicitly annotate individual services. Any service which is not inside a global namespace is considered local even if it has an associated global annotation. Such a service will not share backends with remote clusters in any case.  

Reviewer 3: I mean auto marking/forcing global service in a global namespace. If you define a local namespace it seems fine to not consider the Services marked as global to be actually global because this shouldn't work.

> In the namespace based export mode, all services within global namespaces would automatically be marked as global/local depending on the provided annotation on the namespace. Users are not required to explicitly annotate individual services. Any service which is not inside a global namespace is considered local even if it has an associated global annotation. Such a service will not share backends with remote clusters in any case.  

Author: @MrFreezeex updated. I am also thinking that the functionality to default all non-annotated namespaces to global (inverse case) can be done in the next stage since it was an optimization for users having large number of global namespaces? That'll allow us to deliver the first iteration faster and the optimization can be built on top of the same.

> In the namespace based export mode, all services within global namespaces would automatically be marked as global/local depending on the provided annotation on the namespace. Users are not required to explicitly annotate individual services. Any service which is not inside a global namespace is considered local even if it has an associated global annotation. Such a service will not share backends with remote clusters in any case.  

Author: Keeping that in scope. The CFP is updated with the previously recommended changes. Resolving the comment.

### Summary of Enforcement Logic

| Source Pod in Global NS | Destination Pod in Global NS | Network Policy Applies | Traffic Allowed? |
|-------------------------|-------------------------------|-------------------------|------------------|
| Yes                     | Yes                           | Yes                     | ✅ Enforced normally |
| No                      | Yes                           | Ingress Policy          | ❌ Works for tunnel mode. Does not work for native routing mode  |
| Yes                     | No                            | Egress Policy           | ❌ Works for tunnel mode. Does not work for native routing mode |
| No                      | No                            | Any Policy              | ❌ Works for tunnel mode. Does not work for native routing mode  |
| Any                     | Any                           | No Policies             | ✅ Allowed (no restrictions) |

Even though the policies work for tunnel routing mode, for consistency in customer experience, we do not provide support for network policies for workloads outside global namespaces. 

## Global Service Support

> In the scoped-export mode, all services within global namespaces would automatically be marked as "global". Customers are not required to explicitly annotate individual services. Also, a service which is annotated as global but does not reside in a global namespaces would no longer be global.  

Reviewer 3: I am not sure did you removed details about scoped-export in the ConfirMap but this option is still supposed to exist? If so could you still detail that you are going to add it in the ConfigMap?

### Implementation details

#### Allow/Deny listing namespace
To allowlist a namespace and mark it as global, customers can annotate the namespace with:

```
namespace.cilium.io/global: "true"
```

The ClusterMesh API server will watch namespace resources. Adding this annotation to any namespace will result in the following steps:

- A full synchronization is triggered 
- Only entries associated with namespaces explicitly annotated as global will be exported.
- Entries in etcd that do not belong to any global (allowlisted) namespace will be deleted.
- All unannotated namespaces or entries with denied annotations are implicitly treated as non-global and excluded from export.

To explicitly denylist a namespace, customers can annotate it with:

```
namespace.cilium.io/local: "true"

> ```

Reviewer 3: Could you explain why the two annotation instead of one? For instance in what way `namespace.cilium.io/global: "false"` is different from `namespace.cilium.io/local: "true"`?

Author: They are different due to handling of unannotated namespace. When we have an annotation marking a service as global, the others are by default marked as local. Inversely in the other case, the others are marked as global. We can technically combine them into a single one but I wanted to separate out the annotations due to this reason

Reviewer: I agree with Arthur's comment here. Having the export mode controlled by which annotation is used looks fragile and potentially confusing to me, compared to a straightforward configuration option.

- The system exports the CiliumEndpoints and CiliumIdentities with the updated list of global namespaces
- If no namespaces remain annotated post removal of the annotations, the behavior defaults to the existing behavior of the `clustermesh-apiserver`

#### Network Policies 
An important part of this implementation is the functioning of network policies. In tunnel mode, the identitiy information is embedded in the network packet itself whereas in native routing mode, the identity information at the destination is derieved through the destination cluster's ipcache. As a result, the network policies would not be enforced due to missing identity definition/security labels in native mode and security labels in the tunnel mode. 

#### MCS Support 
MCS-API ServiceExport and ServiceImport CRs will be only supported from a global Namespace. On a local namespace the following will be happening:
- ServiceExport created by a user will report a Condition to signal that the namespace is local and the Service is not actually exported to the clustermesh-apiserver.

> - ServiceImport will still be created by the Cilium Operator but will also report a Condition to signal the user that the Service cannot be imported. Also the derived Service associated to the ServiceImport will not be created and any existing one will be deleted.

Reviewer 3: I think it's fine somewhere in a MCS-API section because it's a behavior exclusively tied to MCS-API/MCS-API controller. We don't even have to do this because the global service don't work as is but it's more for consistency as we know the ServiceImport won't really work thus we don't have to create the derived service more than anything else

> - ServiceImport will still be created by the Cilium Operator but will also report a Condition to signal the user that the Service cannot be imported. Also the derived Service associated to the ServiceImport will not be created and any existing one will be deleted.

Reviewer 3: (I miss-read that you commented all the line rather than only the last one, but yes it's mainly about updating the MCS controllers to add a namespace check and put additional "Conditions" on their status to signal that there's something wrong basically)

> - ServiceImport will still be created by the Cilium Operator but will also report a Condition to signal the user that the Service cannot be imported. Also the derived Service associated to the ServiceImport will not be created and any existing one will be deleted.

Reviewer: Ah, sorry for not being clear. What I meant was just plainly moving the whole section as a subsection of the *Global Service Support* one (or at the same level if preferred). It seems a bit weird that this is here under implementation details, while the other is outside.

#### Export Filtering
If the namespace based export mode is enabled, only the CiliumIdentities and CiliumEndpoints under global namespaces will be exported to the remote clusters. 

#### ServiceExport controller 
Provide warning message when a ServiceExport is triggered for a service residing in local namespace 

#### Supported Resource Types
- CiliumEndpoint(Slice)
- CiliumIdentities
- Global Services
- MCS ServiceExport

### Alternative Approach1

#### Add IPtoSvc and IdentityToIP cache in Clustermesh APIserver 

Since clustermesh-apiserver already has watches setup on CiliumIdentities and CiliumEndpoints, we can maintain mappings of IPtoSvc and IdentityToIP caches which can provide answers to the following questions while processing each service event, ciliumendpoint event or ciliumidentityevent. To join ciliumidentity and ciliumendpoint information, we also need to maintain ipcache of ciliumendpoints  

- Is a CiliumEndpoint fronted by global service?
- Is a CiliumIdentity referenced by a ciliumendpoint fronted by global service?

> The main challenge in this method is requirement to store all ciliumendpoints in memory 

Author: These are alternative approaches which we had in place before coming up with the global namespace approach. I have kept the details at a high level and linked to the doc below. Do you want any specific details to be added in these sections?

> The main challenge in this method is requirement to store all ciliumendpoints in memory 

Reviewer: I think there's been a misunderstanding here. 

My point is that the current text goes directly into implementation details, without providing context on the high level idea of the approach. To make an explicit example, the *"Add IPtoSvc and IdentityToIP cache in Clustermesh APIserver"* title is an implementation detail which would tell nothing to someone who has not read your initial proposal. IMO, a better alternative would be something along the lines of *"Exporting only endpoints referenced by global services*, and then adapting the paragraph to elaborate on that idea.

And similarly for the second alternative approach below.

> Similarly for Global Service implementation, since only the clients under global namespaces are enabled for cross cluster export, all services under `global-namespaces` are global by default and customers cannot create global services under non global(local) namespaces.  

Reviewer 3: nit: there is still two mentions of `global-namespaces` here

Author: Removed it

 
 ### Alternative Approach2

#### Update CiliumEndpoint, CiliumIdentity and CiliumEndpointSlice controllers 
Update the CRD controllers to internally annotate CiliumEndpoints, CiliumIdentities and CiliumEndpointSlices associated with 
- Pods that are annotated by the user 
- Endpoints behind global services 

> Once the annotations are added in the CRD controllers in Cilium Agent/Operator, the Clustermesh APIserver can filter out the non annotated CRDs and prevent them from being exported

Reviewer: Same here.

#### MCS Support 
MCS-API ServiceExport and ServiceImport CRs will be only supported from a global Namespace. On a local namespace the following will be happening:
- ServiceExport created by a user will report a Condition to signal that the namespace is local and the Service is not actually exported to the clustermesh-apiserver.
- ServiceImport will still be created by the Cilium Operator but will also report a Condition to signal the user that the Service cannot be imported. Also the derived Service associated to the ServiceImport will not be created and any existing one will be deleted.

### ClusterMesh API Server Changes

#### Process annotations 
We will add new watch on namespaces to check for global or local annotations and maintain internal sets for the same. Once we have the list, we will trigger a full sync operation to update the export based on the provided annotation. 

#### Export Filtering
If the namespace based export mode is enabled, only the CiliumIdentities and CiliumEndpoints under global namespaces will be exported to the remote clusters. 

> #### ServiceExport controller 
> Provide warning message when a ServiceExport is triggered for a service residing in local namespace 
> 

Reviewer 3: I think you can remove that since we are talking about that in the MCS support above and this is not even done in the clustermesh api server that we are setting conditions (it's in the operator)
```suggestion
#### ServiceExport controller 
Provide warning message when a ServiceExport is triggered for a service residing in local namespace 

```

Reviewer 3: You would be necessarily doing both because it's the same code called on both actually

Reviewer 3: But I don't think you have to mention that precisely into the CFP or at least not in dedicated sections... I was only commenting about complexing this text not the actual implementation

## Motivation

The existing Cilium Clustermesh implementation distributes all endpoints and identities across connected clusters, creating scaling constraints as resource volumes grow. This comprehensive propagation model becomes a limiting factor when managing extensive endpoint and identity collections. Performance profiling analysis demonstrates the severity of this scaling challenge: in a 200-cluster Clustermesh deployment with the following configuration parameters, each Cilium agent consumes 15GiB of memory:

- cm-apiserver-replicas=2
- clusters=200
- nodes/cluster=300
- identities/cluster=500
- identities-qps=0.2
- endpoints/cluster=15000
- endpoints-qps=1
- services=0
- services-qps=0

> By restricting export to only resources under global namespaces, we can achieve substantial scalability improvements and operational efficiency gains. The gains in compute from the proposed optimization would also ensure that customers can run multiple replicas of the Clustermesh APIserver which is recommended for production workloads. An important consideration for using this new "scoped-export" mode is that all resources under the global namespaces are exported to clusters backend. This implies that the memory and compute optimizations might be suboptimal if the cluster has most of the resources under the global namespaces. 

Author: Removed it

| Yes                     | No                            | Egress Policy           | ❌ Works for tunnel mode. Does not work for native routing mode |
| No                      | No                            | Any Policy              | ❌ Works for tunnel mode. Does not work for native routing mode  |
| Any                     | Any                           | No Policies             | ✅ Allowed (no restrictions) |

Even though the policies work for tunnel routing mode, for consistency in customer experience, we do not provide support for network policies for workloads outside global namespaces. 

## Global Service Support
In the scoped-export mode, all services within global namespaces would automatically be marked as "global". Customers are not required to explicitly annotate individual services. Also, a service which is annotated as global but does not reside in a global namespaces would no longer be global.  

## MCS support
Similar to global services, MCS support is only available for services created under global namespaces. 

### Implementation details

#### Allow/Deny listing namespace

> To allowlist a namespace and mark it as global, customers can annotate the namespace with:

Reviewer: I think we may want to discuss a bit whether it should be an annotation or a label. While I agree that an annotation would be consistent with the service one, I wouldn't automatically exclude the usage of a label to easily allow users to list all global namespaces (opposed to needing an extra CLI tool).

Author: If there's a strong preference for either, we can change it. The thinking process behind keeping it annotation was to keep the behavior consistent with global services

```
namespace.cilium.io/global: "true"
```

The ClusterMesh API server will watch namespace resources. Adding this annotation to any namespace will result in the following steps:

> - A full synchronization is triggered 

Reviewer: With the operational mode controlled by a configuration, the full synchronization should be triggered only for the resources inside the namespace in which the annotation got changed (if and only if the annotation got actually changed).

#### MCS API 

> We would not crete ServiceExport CRDs for local namespaces. This would inturn prevent the corresponding ServiceImport CRDs from being created in the remote clusters. Since we support transition to scoped-export mode without restarting the pods, it is possible that the CRDs might have been created before the customers annotate their namespaces. We would not provide support for the functionality in local namespaces even if the CRDs exist.  

Reviewer: > ServiceExport are user controlled (see https://docs.cilium.io/en/latest/network/clustermesh/mcsapi/#exporting-a-service), so as I was suggesting in my comment we should rather probably signal to the user that the export does not work.

Absolutely, I saw that but then forgot to comment in this first round. We could add a condition for that, right?

> Also I don't know what we should do about ServiceImport, if we just block ServiceExport in the local cluster it could technically still be created if we are just aiming to block exporting. This probably is the same question as a global service in a namespace that does not export resource: should it get backends from remote clusters or do we consider like it was not even annotated/global?

I'd probably treat all non-global namespaces as non-existing from the MCS_API point of view, hence not creating a ServiceImport even if a `ServiceExport` is created in one of the remote clusters (and the namespace is global there). This is one additional case in which using a label instead of an annotation would make things super simple (basically filtering at the informer level, so that we see a create/delete event if the label gets added/removed).

> We would not crete ServiceExport CRDs for local namespaces. This would inturn prevent the corresponding ServiceImport CRDs from being created in the remote clusters. Since we support transition to scoped-export mode without restarting the pods, it is possible that the CRDs might have been created before the customers annotate their namespaces. We would not provide support for the functionality in local namespaces even if the CRDs exist.  

Reviewer: >> I'd probably treat all non-global namespaces as non-existing from the MCS_API point of view, hence not creating a ServiceImport even if a ServiceExport is created in one of the remote clusters (and the namespace is global there).
>
> This depends if we keep the naming of "global namespace" or if we go to a naming only related to "don't export resources from this namespace" IMO. 

Good point on the global vs export (shared) distinction. The current naming implies the first, but the other approach may work as well (at first sight it seems a bit less intuitive to me, but I need to think a bit more about it). Anyways, the intended behavior should be be clarified as part of the CFP. 

> Also if we do that, regular service annotated with the global service annotation should probably be treated like if they didn't have the global service annotation I guess?

Yep, absolutely, the global services and MCS-API behavior should be consistent.

>> This is one additional case in which using a label instead of an annotation would make things super simple (basically filtering at the informer level, so that we see a create/delete event if the label gets added/removed).
>
> It's slightly easier indeed but I don't know if I am fully convinced by labels yet as there's to my knowledge not much else that change some behavior like that based on labels on Kubernetes/Cilium besides defining some extra resource that have some label selectors.

I agree, I'm not fully convinced either, especially in terms of consistency. Just mentioning that it has some advantages.

> We would not crete ServiceExport CRDs for local namespaces. This would inturn prevent the corresponding ServiceImport CRDs from being created in the remote clusters. Since we support transition to scoped-export mode without restarting the pods, it is possible that the CRDs might have been created before the customers annotate their namespaces. We would not provide support for the functionality in local namespaces even if the CRDs exist.  

Reviewer 3: @krunaljain this thread is not addressed, ServiceExport and ServiceImport should not get the same treatment since ServiceExport are use controller and ServiceImport are created by the Cilium operator.

> We would not crete ServiceExport CRDs for local namespaces. This would inturn prevent the corresponding ServiceImport CRDs from being created in the remote clusters. Since we support transition to scoped-export mode without restarting the pods, it is possible that the CRDs might have been created before the customers annotate their namespaces. We would not provide support for the functionality in local namespaces even if the CRDs exist.  

Reviewer 3: Assuming we can introduce conditions on ServiceImport (which I think is likely, PR for that is here fyi https://github.com/kubernetes-sigs/mcs-api/pull/112), adding a failed condition and not creating/removing the related derived service seems right

> We would not crete ServiceExport CRDs for local namespaces. This would inturn prevent the corresponding ServiceImport CRDs from being created in the remote clusters. Since we support transition to scoped-export mode without restarting the pods, it is possible that the CRDs might have been created before the customers annotate their namespaces. We would not provide support for the functionality in local namespaces even if the CRDs exist.  

Reviewer 3: I meant ServiceImport but I sent the wrong PR, here is the right one: https://github.com/kubernetes-sigs/mcs-api/pull/113

> We would not crete ServiceExport CRDs for local namespaces. This would inturn prevent the corresponding ServiceImport CRDs from being created in the remote clusters. Since we support transition to scoped-export mode without restarting the pods, it is possible that the CRDs might have been created before the customers annotate their namespaces. We would not provide support for the functionality in local namespaces even if the CRDs exist.  

Reviewer 3: And if we cannot add a condition I would vote for not creating any ServiceImport in a "local" namespace (and removing those still there).

ServiceExport as there are conditions available and it is user controller we should let them be created and add a condition signaling something wrong (as we are also changing the condition type in MCS-API you don't have to be necessarily specific but with the current state it would be "Valid" and potentially "Accepted" with ServiceExport conditions PR current state).

### Summary of Enforcement Logic

| Source Pod in Global NS | Destination Pod in Global NS | Traffic Allowed? |
|-------------------------|-------------------------------|------------------|
| Yes                     | Yes                                                | ✅ Enforced normally |
| No                      | Yes                                                | ❌ Not Supported  |
| Yes                     | No                                | ❌ Not Supported |
| No                      | No                                       | ❌ Not Supported  |
| Any                     | Any                                        | ✅ Allowed (no restrictions) |

## Global Service Support
For a service to be considered a Global Service, it must satisfy the following conditions 
- It must be annotated with existing global service annotation ***service.cilium.io/global: "true"***
- It must reside within a global namespace 

> In the above table, Not Supported implies Cilium cannot enforce any network policies—neither allow nor deny rules are evaluated or applied across clusters. For Supported usecases, the behavior of Network Policies remains exactly the same as is in the existing Cilium versions. 

Reviewer: > In the above table, Not Supported implies Cilium cannot enforce any network policies—neither allow nor deny rules are evaluated or applied across clusters.

This is still not specific enough IMO. What does *"rules are evaluated or applied across clusters"* mean practically? Is all traffic allowed (which would sound fairly scary to me from a security standpoint)? Is all traffic blocked? Would traffic be potentially misclassified?

#### Workflow

1. **User Annotation**: 
   - Users annotate certain pods with a predefined key (e.g., `clustermesh.cilium.io/global-service-client=true`) to indicate that the pod is a client/endpoint of one or more global services.

2. **CRD Controller Processing**:
   - The Cilium Agent and Operator monitors pod changes.
   - When a pod with the client annotation is detected, the associated `CiliumEndpoint`, `CiliumIdentity`, and `CiliumEndpointSlice` are automatically annotated internally.

3. **Purpose of Internal Annotations**:
   - These internal annotations allow downstream components, such as the `clustermesh-apiserver`, to easily identify and filter relevant resources during export.
   - This avoids the overhead of recomputing client-pod status or performing expensive label queries in the API server.

> The main complexity of this approach lies in inferring global service endpoints internally along with annotated client pods. The client annotations are processed in the CiliumEndpoint, CiliumIdentity and CiliumEndpointSlice CRD controllers which would lead to changes in all the existing Cilium components (Agent, Operator and ClusterMesh APIserver). 

Author: It is common to both the approaches since they both require processing the pod annotations in the CRD controllers

# CFP-39876: Namespace based Export Control for ClusterMesh

**SIG:** SIG-ClusterMesh

**Sharing:** Public

**Begin Design Discussion:** 2025-06-02

**End Design Discussion:**

**Cilium Release:** 1.19

> **Authors:** Krunal Jain <krunaljain@microsoft.com>, Vamsi Kalapala <Krishna.Vamsi@microsoft.com>

Reviewer 4: ```suggestion
**Authors:** Krunal Jain <krunaljain@microsoft.com>, Vamsi Kalapala <Krishna.Vamsi@microsoft.com>

**Status:** Implementable
```

## Goals

- Enhance Cilium ClusterMesh scalability through selective endpoint and identity distribution
- Full network policy support for pods inside global namespaces
- Cross-cluster pod to pod connectivity inside global namespaces
- Global service and MCS-API support for resources under global namespaces

## Non-Goals

- Network policies for non global(local) namespaces
- Cross-cluster pod to pod connectivity for non global(local) namespaces
- Global service and MCS-API support for non global(local) namespaces
- Support Global/Local namespaces for non CRD modes of identity allocation 
 
## Network Policy 
### Cross-Cluster Communication and Network Policy Behavior

> Full support for **cross-cluster communication** and **network policy enforcement** is available **only when both the source (client) and destination (server) pods**—which may optionally be accessed via a service—**are part of a _global namespace_**. Network policies are not enforced if either the source or destination pod is in a **non-global namespace**.

Reviewer 5: Network policies are always enforced. The question is the information that is used to enforce the policies.

This suggestion would drop the incorrect statement:

```suggestion
Full support for **cross-cluster communication** and **network policy enforcement** is available **only when both the source (client) and destination (server) pods**—which may optionally be accessed via a service—**are part of a _global namespace_**.
```

Alternatively you could add a statement explaining more clearly how network policies for local namespaces will not take into account labels for local namespaces in other clusters.

#### Why Current Policies Can't Work

Cilium `NetworkPolicy` typically uses selectors like:

```yaml
endpointSelector:
  matchLabels:
    app: frontend
ingress:
  - fromEndpoints:
    - matchLabels:
        app: backend
        tier: api
```

Without labels, the policy engine cannot:
-Match fromEndpoints selectors
-Evaluate matchLabels conditions
-Distinguish between different application identities

When using this feature, cross-cluster communication requires both the source and destination namespaces to be global, and both clusters to have an established cluster mesh. The evaluation of network policy for different cases is summarized below:

### Summary of Enforcement Logic

| Source Pod in Global NS | Destination Pod in Global NS | Can you allow traffic based on Labels? |
|-------------------------|------------------------------|------------------|
| Yes                     | Yes                          | ✅ Enforced normally |
| No                      | Yes                          | ❌ Not Supported  |
| Yes                     | No                           | ❌ Not Supported |
| No                      | No                           | ❌ Not Supported  |

> In the above table, Not Supported implies the policies are not enforced due to missing source identity or labels information at the destination cluster. 

Reviewer 5: Wording around "policies are not enforced" can be misinterpreted to imply that network policy enforcement will be completely disabled, but that is not what the CFP proposes.

```suggestion
In the above table, Not Supported implies that policy enforcement will not take into account labels on Pods in local namespaces in a peer cluster. 
```

We might clarify further that this means that even if a policy explicitly allows traffic for a remote cluster for pods with specific labels, that policy may not allow the traffic if the remote Pod's namespace is local. 

I think we're on the same page here, but there is a subtlety in the wording between "policy enforcement has surprising behavior" vs. "there is no policy enforcement".

- clustermesh-default-global-namespace (true/false) - Marks implicitly all non annotated namespaces as global if set to true. Marks implicitly all non annotated namespaces as local if set to false. This option is set to true by default to preserve the current behavior.

#### Allow/Deny listing namespace
To allowlist a namespace and mark it as global, users can annotate the namespace with:

```
clustermesh.cilium.io/global: "true"
```

The ClusterMesh API server will watch namespace resources. Adding, changing or removing this namespace annotation will result in the following steps:

- A full synchronization is triggered for all resources within the newly annotated namespaces 
- Only entries associated with namespaces explicitly annotated as global will be exported.
- Unannotated namespaces are treated as global/local depending on the value of config `clustermesh-default-global-namespace`

To explicitly mark a namespace as local, the annotation can be set as:

```
clustermesh.cilium.io/global: "false"
```

When the annotation is removed or disabled for a namespace:

> - A full synchronization is triggered to mark all resources within the annotated namespace as global/local

Reviewer 5: What is involved in this "mark all resources" step?

When it comes to user-managed resources I think it makes a lot of sense to ensure the user is responsible for setting the annotations. If there are resources managed only by Cilium, I think it makes sense for Cilium to propagate such annotations. What I'm less clear about here is for instance does this mean that if the user adds a namespace annotation, Cilium will perform write operations against Pods and Services in the namespace to propagate the annotations? Is that what this statement means? (If yes, is that really necessary?)

Reviewer 5: Ah right, this is about the forward propagation into etcd :+1: Makes sense yep.

- ServiceExport created by a user will report a Condition to signal that the namespace is local and the Service is not actually exported to the clustermesh-apiserver.
- ServiceImport will still be created by the Cilium Operator but will also report a Condition to signal the user that the Service cannot be imported. Also the derived Service associated to the ServiceImport will not be created and any existing one will be deleted.

### ClusterMesh API Server Changes

#### Process annotations 
We will add new watch on namespaces to check for global or local annotations and maintain internal sets for the same. Once we have the list, we will trigger a full sync operation to update the export based on the provided annotation. 

#### Export Filtering
Only the CiliumIdentities and CiliumEndpoints under global namespaces will be exported to the remote clusters. 

#### ServiceExport controller 
Provide warning message when a ServiceExport is triggered for a service residing in local namespace 

#### Existing Cilium Resource Types
- CiliumEndpoint(Slice)
- CiliumIdentities
- Global Services
- MCS ServiceExport

## Impacts / Key Questions
---

### Impact: Breakage during upgrade

> - The changes are expected to be backward compatible in the sense that the new scoped export logic only takes effect once the user has explicitly annotated namespaces as global or local. As a result, existing clusters will continue to function as before until the user opts into the new behavior. The `clustermesh-default-global-namespace` config option will determine the default behavior for unannotated namespaces.

Reviewer 6: today, all identities are shared across all clusters. When user upgrades to this new version, that means upgrading to new agent and apiserver, existing workloads with network policies across clusters continue to work without annotating anything? and only new deployments requires annotation?

Author: The annotation is not required in any case. It acts as a feature flag for users to opt into this scoped-export mode. Without annotation, the behavior of ClusterMesh would be exactly the same as it is today

---

### Impact: Debugging multi-cluster environment connectivity may become more difficult
- Failures due to missing identity or label exports may be non-obvious and require deeper inspection into the global/local namespace configuration.
- Cross-cluster connectivity issues will now depend not only on service presence or routes but also on namespace-level annotations and default behaviors requiring additional verifications on the value of config `clustermesh-default-global-namespace` along with the actual annotations present on the namespaces during debugging.
- If an application is sending traffic to a destination in another cluster and the traffic is dropped due to policy (as seen in hubble observe), users should inspect whether the source and destination namespaces are marked as global.
---
### Alternative Approach 1

### Exporting CiliumEndpoint and CiliumIdentity for Global Services

> By default, all endpoints fronted by **global services** are expected to have their `CiliumEndpoint` and `CiliumIdentity` information shared across clusters to support cross-cluster load balancing. If we can determine whether a given `CiliumEndpoint` or `CiliumIdentity` is fronted by a global service, we can use that as the basis for deciding whether to export it.

Reviewer 6: so ipcache is not fully exported and is derived based on exported ciliumendpoint and identity?

Author: IPcache is always derived from CiliumEndpoint objects. The cache now will just have global service endpoints instead of all CiliumEndpoints

To enable clients of global services to communicate with their backend service endpoints, we also need a mechanism for users to explicitly mark certain pods as **clients** of global services. This can be achieved through a pod-level annotation.

Since the `clustermesh-apiserver` already watches `CiliumEndpoints` and `CiliumIdentities`, we can maintain in-memory caches to support export decisions:

> - **`IPToService`**: a mapping from IP addresses to services that front them.

Reviewer 6: is this backend ip to cluster/lb ip?

Author: this is the mapping of 

CiliumEndpoint IP ----> List of Services fronting that front that CiliumEndpoint

- **`IdentityToIP`**: a mapping from identities to the set of IPs they represent.

> - **`IPCache`**: to associate IPs back to `CiliumEndpoints`.

Reviewer 6: i guess this is different from cilium ipcache?

# CFP-39876: Namespace based Export Control for ClusterMesh

**SIG:** SIG-ClusterMesh

**Sharing:** Public

**Begin Design Discussion:** 2025-06-02

**End Design Discussion:**

**Cilium Release:** 1.19

**Authors:** Krunal Jain <krunaljain@microsoft.com>, Vamsi Kalapala <Krishna.Vamsi@microsoft.com>

## Summary

> This CFP introduces selective distribution of CiliumEndpoint/CiliumEndpointSlice and CiliumIdentity information via Cilium ClusterMesh by adding a new export mechanism for the ClusterMesh API server, which restricts cross-cluster propagation only to resources that reside in a set of allowlisted namespaces, also known as _global namespaces_. Users can mark a namespace as global through a dedicated annotation applied to the namespace itself. This proposal targets improved scalability while acknowledging that cross-cluster endpoint access and network policy enforcement is only enabled for pods outside global namespaces.

Reviewer: Good catch, this should be actually *inside global namespaces*

> This CFP introduces selective distribution of CiliumEndpoint/CiliumEndpointSlice and CiliumIdentity information via Cilium ClusterMesh by adding a new export mechanism for the ClusterMesh API server, which restricts cross-cluster propagation only to resources that reside in a set of allowlisted namespaces, also known as _global namespaces_. Users can mark a namespace as global through a dedicated annotation applied to the namespace itself. This proposal targets improved scalability while acknowledging that cross-cluster endpoint access and network policy enforcement is only enabled for pods outside global namespaces.

Author: Fixed it. Thanks

- clusters=200
- nodes/cluster=300
- identities/cluster=500
- endpoints/cluster=15000
- services/cluster=0

By restricting export to only resources under global namespaces, we can achieve substantial scalability improvements and operational efficiency gains. An important consideration for using this new global/local namespace feature is that all resources under the global namespaces are exported to the remote clusters. This implies that the memory and compute optimizations might be suboptimal if the cluster has most of the resources under the global namespaces. Without the proposed optimizations, each clustermesh-apiserver replica would export the entire local state to remote clusters.  

## Goals

- Enhance Cilium ClusterMesh scalability through selective endpoint and identity distribution
- Full network policy support for pods inside global namespaces
- Cross-cluster pod to pod connectivity inside global namespaces
- Global service and MCS-API support for resources under global namespaces

## Non-Goals

- Network policies for non global(local) namespaces
- Cross-cluster pod to pod connectivity for non global(local) namespaces

> - Global service and MCS-API support for non global(local) namespaces

Reviewer 5: That's fair yeah. Maybe this proposed thread is just reading too deep into the text. It should be fine to leave as-is.

- Support Global/Local namespaces for non CRD modes of identity allocation 
 
## Network Policy 
### Cross-Cluster Communication and Network Policy Behavior

Full support for **cross-cluster communication** and **network policy enforcement** is available **only when both the source (client) and destination (server) pods**—which may optionally be accessed via a service—**are part of a _global namespace_**.

### Behavior When Pods Are Not in Global Namespaces
Network policies in Cilium rely heavily on **identity labels** for policy matching. When the identity along with associated labels is not exported, the following behavior is observed:

#### Tunnel Routing Mode

- Identity ID is exported ✅  
- Labels are NOT exported ❌  

> - **Result**: Destination cluster knows _who_ the source is (by ID) but not _what_ it is (labels). The policy applied through the labels is not enforced. 

Author: @joestringer this can be implemented although will require changes in the underlying bpf programs to consume namespace events. If there is a marked global namespace, then the logic needs to be executed. It might be a bit complex to achieve the same. Hence, directly marked the use case as not supported for v1. Can this be a future goal?

> - **Result**: Destination cluster knows _who_ the source is (by ID) but not _what_ it is (labels). The policy applied through the labels is not enforced. 

Reviewer: > This is assuming the source cluster knows how to route this specific Pod's traffic to the destination cluster (presumably via the specific Node the destination resides on?). Is that a given?

> For the case where the destination is in a local namespace in a remote cluster, the traffic wouldn't even get routed to the destination, so the point is somewhat moot for that case.

AFAIK, routing strictly speaking would work in the majority of cases, as either the clusters are configured in native routing mode (or ENI), and in that case the network knows how to forward the packets, or it operates in tunnel mode, and in that case it is likely that the IPAM is node-based, and there's a fallback ipcache entry handling the routing to the CIDR associated with that node.

> Throwing some ideas out here, but if the source cluster knows that the source Identity is in a local namespace and the traffic is destined to a Pod in a remote cluster in a global namespace, maybe the source cluster should just drop with a new reason something like DROPPED (Local Pod unroutable on Cluster Mesh)

This should be relatively easy to achieve whenever there's a per-node CIDR, as it would be basically a matter of assigning a different identity to the fallback ipcache entry (currently it is world for backwards compatibility) for the nodes in remote clusters. Ingress policies in tunnel mode would require an extra ipcache lookup, though, as it is not possible to rely on the source identity conveyed via the tunnel metadata (which would be the real ID, not this fallback).

If there's no per-node CIDR, users would need to configure an entry that covers the entire cluster CIDR, which can get tricky in cloud environments if the VPC is reused for e.g., multiple clusters (I don't have much experience there on the typical configurations though).

Overall, I agree with @krunaljain that this is something for a follow-up step though.

> - **Result**: Destination cluster knows _who_ the source is (by ID) but not _what_ it is (labels). The policy applied through the labels is not enforced. 

Reviewer: Yep, that would be definitely better :+1:

#### Native Routing Mode

- Identity not present in ipcache ❌  
- Identity defaults to "world" identity ❌  
- Labels are NOT exported ❌  

> - **Result**: All cross-cluster traffic appears as generic **"world"** traffic. The identity and its associated labels are lost and no associated policies can be enforced.

Reviewer: I agree with @joestringer that the description here should be changed to a more general statement, without tying it to the routing mode, as otherwise we bind ourselves too much to the current implementation, which is susceptible to changes.

My proposal would be something along the lines of:

> Cross-cluster traffic from a source pod not part of a global namespace to a destination pod matched by an ingress policy is dropped, unless from-world traffic is explicitly allowed by the policy, in which case it *may* be allowed, depending on the specific Cilium configuration. Similarly, cross-cluster traffic from a source pod matched by an egress policy to a destination pod not part of a global namespace is dropped, unless to-world traffic is explicitly allowed by the policy (either globally, or via a specific CIDR/FQDN rule), in which case it is allowed.

> - **Result**: All cross-cluster traffic appears as generic **"world"** traffic. The identity and its associated labels are lost and no associated policies can be enforced.

Reviewer: The main caveat is that this applies now that the fallback ipcache entries for the node CIDRs are associated with the `world` identity, but it may change in the future if that gets changed (as e.g., proposed in the comment above).

I also wonder how much emphases we want to put on the fact that CIDR policies may be used to allow the traffic in this context, as that risks being fairly confusing for users. For instance, they may allow certain traffic via a CIDR policy, and then lose access to it once they mark the namespace as global, given that the specific endpoint identity takes precedence.

> - **Result**: All cross-cluster traffic appears as generic **"world"** traffic. The identity and its associated labels are lost and no associated policies can be enforced.

Reviewer: > probably for user instructions we would suggest that if you want to apply network policies for cross-cluster traffic, users should just mark both namespaces as global

Yep, totally agree. I'd not mention CIDR policies in user-documentation in this context, to not introduce extra sources of confusion.

> - **Result**: All cross-cluster traffic appears as generic **"world"** traffic. The identity and its associated labels are lost and no associated policies can be enforced.

Author: Dropped references to routing modes and made the section generic

---

#### Why Current Policies Can't Work

Cilium `NetworkPolicy` typically uses selectors like:

```yaml
endpointSelector:
  matchLabels:
    app: frontend
ingress:
  - fromEndpoints:
    - matchLabels:
        app: backend
        tier: api
```

Without labels, the policy engine cannot:
-Match fromEndpoints selectors
-Evaluate matchLabels conditions
-Distinguish between different application identities

> If either the source or the destination pod does **not** belong to a global namespace, cross-cluster communication **may still succeed** under implementation-dependent conditions. However, this communication flow is not officially supported, and may be subject to changes at any time. The evaluation of network policy for different cases is summarized below:

Author: @joestringer the network policy is tied to the routing mode in the existing Cilium setup. Hence, the enforcement of the network policy is different for different routing modes. As a result, I have added the expected behavior with different routing modes to explain the end state of policy enforcement.

> If either the source or the destination pod does **not** belong to a global namespace, cross-cluster communication **may still succeed** under implementation-dependent conditions. However, this communication flow is not officially supported, and may be subject to changes at any time. The evaluation of network policy for different cases is summarized below:

Reviewer 5: I don't see it that way. I see it as: Each network enforcement point makes a determination of the source and peer identities and then enforces the policy. The different routing modes may choose to propagate the identity information in-band or out-of-band, but the enforcement is still made based on the understanding of the peers at the enforcement point. Assuming that the traffic passes the enforcement point (even if that traffic is "misclassified" in a global sense but that classification matches the policy), _then_ it's up to routing to either provide connectivity to the destination or not.

### Summary of Enforcement Logic

| Source Pod in Global NS | Destination Pod in Global NS | Traffic Allowed? |
|-------------------------|------------------------------|------------------|
| Yes                     | Yes                          | ✅ Enforced normally |
| No                      | Yes                          | ❌ Not Supported  |
| Yes                     | No                           | ❌ Not Supported |
| No                      | No                           | ❌ Not Supported  |

> In the above table, Not Supported implies the policies are not enforced due to missing source identity or labels information at the destination cluster. The end state of the policy enforcement is dependent on the routing mode of the cluster as discussed above. 

Author: @joestringer the context around routing modes is required since the network policies behave differently with different routing modes. For tunnel mode, the identity id is passed in packet header whereas the native routing mode extracts the id from the ipcache. The labels are not passed in either of these cases. Hence, the mention of routing modes in this section

> In the above table, Not Supported implies the policies are not enforced due to missing source identity or labels information at the destination cluster. The end state of the policy enforcement is dependent on the routing mode of the cluster as discussed above. 

Reviewer: I agree with Joe that the sentence could be reformulated to avoid tying it to the routing mode (c.f. the comment above).

> In the above table, Not Supported implies the policies are not enforced due to missing source identity or labels information at the destination cluster. The end state of the policy enforcement is dependent on the routing mode of the cluster as discussed above. 

Author: Dropped the sentence

- ServiceExport created by a user will report a Condition to signal that the namespace is local and the Service is not actually exported to the clustermesh-apiserver.
- ServiceImport will still be created by the Cilium Operator but will also report a Condition to signal the user that the Service cannot be imported. Also the derived Service associated to the ServiceImport will not be created and any existing one will be deleted.

### ClusterMesh API Server Changes

#### Process annotations 
We will add new watch on namespaces to check for global or local annotations and maintain internal sets for the same. Once we have the list, we will trigger a full sync operation to update the export based on the provided annotation. 

#### Export Filtering
Only the CiliumIdentities and CiliumEndpoints under global namespaces will be exported to the remote clusters. 

#### ServiceExport controller 
Provide warning message when a ServiceExport is triggered for a service residing in local namespace 

#### Existing Cilium Resource Types
- CiliumEndpoint(Slice)
- CiliumIdentities
- Global Services

> - MCS ServiceExport

Reviewer 5: Ah yes sharing only security identities would only really work for ingress policy in tunneling mode (with an additional caveat of delegating high trust to the source cluster). Doesn't help with egress policy.

There might be other ideas there like if you had only security identities for namespaces, and had an IPAM that set up dedicated ranges per namespace, maybe you could reduce the sheer number of IP ranges to share... but that in itself has a few caveats and assumptions.

I agree with the perspective that a smarter filtering angle looks a bit too complex to pursue for now.

> 

Reviewer 5: To be clear on this feedback, the lines starting with `#` are my suggestions, and the lines in the italic are copied directly from the template to instruct you about how to fill out the sections. What I'd like to see here is some consideration of the upgrade and operational impacts, and the key question around defaults. We can of course discuss this too. If you have a sense of what might be important to discuss/communicate already, I'd encourage you to fill out this suggestion in the CFP and we can discuss the text subsequently. If you'd like to discuss one of these aspects more first, feel free to just ask in the thread here.

> 

Author: Added details. Let me know if you need additional details

- Cross-cluster pod to pod connectivity inside global namespaces
- Global service and MCS-API support for resources under global namespaces

## Non-Goals

- Network policies for non global(local) namespaces
- Cross-cluster pod to pod connectivity for non global(local) namespaces
- Global service and MCS-API support for non global(local) namespaces
- Support Global/Local namespaces for non CRD modes of identity allocation 
 
## Network Policy 
### Cross-Cluster Communication and Network Policy Behavior

Full support for **cross-cluster communication** and **network policy enforcement** is available **only when both the source (client) and destination (server) pods**—which may optionally be accessed via a service—**are part of a _global namespace_**. Network policies are not enforced if either the source or destination pod is in a **non-global namespace**.

### Behavior When Pods Are Not in Global Namespaces
Network policies in Cilium rely heavily on **identity labels** for policy matching. When the identity along with associated labels is not exported, the following behavior is observed:

#### Tunnel Routing Mode

- Identity ID is exported ✅  
- Labels are NOT exported ❌  

> - **Result**: Destination cluster knows _who_ the source is (by ID) but not _what_ it is (labels). The policy applied through the labels is not enforced. 

Reviewer: FYI, this is not correct, as it only applies to ingress policies. As for egress policies, the source cluster would not know the identity of the destination pod, neither the ID, nor the labels.

- ServiceExport created by a user will report a Condition to signal that the namespace is local and the Service is not actually exported to the clustermesh-apiserver.
- ServiceImport will still be created by the Cilium Operator but will also report a Condition to signal the user that the Service cannot be imported. Also the derived Service associated to the ServiceImport will not be created and any existing one will be deleted.

### ClusterMesh API Server Changes

#### Process annotations 
We will add new watch on namespaces to check for global or local annotations and maintain internal sets for the same. Once we have the list, we will trigger a full sync operation to update the export based on the provided annotation. 

#### Export Filtering
Only the CiliumIdentities and CiliumEndpoints under global namespaces will be exported to the remote clusters. 

#### ServiceExport controller 
Provide warning message when a ServiceExport is triggered for a service residing in local namespace 

#### Existing Cilium Resource Types
- CiliumEndpoint(Slice)
- CiliumIdentities
- Global Services
- MCS ServiceExport

## Impacts / Key Questions
---

### Impact: Breakage during upgrade

> - The changes are expected to be backward compatible in the sense that the new scoped export logic only takes effect once the user has explicitly annotated namespaces as global or local. As a result, existing clusters will continue to function as before until the user opts into the new behavior.

Reviewer: > I'd be inclined to advocate for a global flag for clustermesh-apiserver that sets the default sharing policy. That way the administrator could opt in/out based on what they expect to share. You could imagine a default "share everything" stance, then the administrator can label all the namespaces the way they want, then modify the default stance to defer to the annotations (and not share everything).

Yes, this is what we agreed upon, and that is described in the rest of the CFP. I agree that the phrasing here is confusing and should be updated, as it seems to refer to the previously proposed approach that we had already discarded for the very same reasons that you are mentioning.

> - The changes are expected to be backward compatible in the sense that the new scoped export logic only takes effect once the user has explicitly annotated namespaces as global or local. As a result, existing clusters will continue to function as before until the user opts into the new behavior.

Author: Added reference to the config

---

### Impact: Debugging multi-cluster environment connectivity may become more difficult
- Failures due to missing identity or label exports may be non-obvious and require deeper inspection into the global/local namespace configuration.

> - For the initial rollout, network policies are not supported if either of the namespaces containing the endpoints are local. This can be a quick debugging step to debug network policy enforcement issues.

Author: Removed
